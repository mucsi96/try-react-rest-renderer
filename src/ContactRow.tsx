import { FC } from "react";

export const ContactRow: FC<{ children: string }> = ({ children }) => (
  <div>{children}</div>
);
