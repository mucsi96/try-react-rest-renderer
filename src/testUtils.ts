import { ElementType, ReactElement } from "react";
import {
  act,
  create,
  ReactTestInstance,
  ReactTestRenderer,
  ReactTestRendererJSON,
  ReactTestRendererNode,
} from "react-test-renderer";
import { expect } from "vitest";

export const waitFor = (
  callback: Function,
  { interval = 50, timeout = 1000 } = {}
): Promise<undefined> =>
  act(
    () =>
      new Promise((resolve, reject) => {
        const startTime = Date.now();

        const nextInterval = () => {
          setTimeout(() => {
            try {
              callback();
              resolve();
            } catch (err) {
              if (Date.now() - startTime > timeout) {
                reject(new Error("Timed out."));
              } else {
                nextInterval();
              }
            }
          }, interval);
        };

        nextInterval();
      })
  );

const serializer = {
  test(value: ReactWrapper): boolean {
    return value instanceof ReactWrapper;
  },
  serialize(val: ReactWrapper, config: any): string {
    const json = val.toJSON();
    const plugin = config.plugins.find(
      (p: any) => p != serializer && p.test(json)
    );
    return plugin.serialize(json, ...[...arguments].slice(1));
  },
};

expect.addSnapshotSerializer(serializer);

function getComponentNameFromType(type: ElementType): string {
  if (typeof type === "function") {
    return (type as any).displayName || type.name || 'null';
  }

  if (typeof type === "string") {
    return type;
  }
  return 'null';
}

function toJSON(
  inst: ReactTestInstance | string
): ReactTestRendererNode | null {
  if (typeof inst === "string") {
    return inst;
  }

  const { children, ...props } = inst.props;
  let renderedChildren = null;
  if (inst.children && inst.children.length) {
    for (let i = 0; i < inst.children.length; i++) {
      const renderedChild = toJSON(inst.children[i]);
      if (renderedChild !== null) {
        if (renderedChildren === null) {
          renderedChildren = [renderedChild];
        } else {
          renderedChildren.push(renderedChild);
        }
      }
    }
  }
  const json: ReactTestRendererJSON = {
    type: getComponentNameFromType(inst.type),
    props: props,
    children: renderedChildren,
  };
  Object.defineProperty(json, "$$typeof", {
    value: Symbol.for("react.test.json"),
  });
  return json;
}

class ReactWrapper {
  private rootRenderer: ReactTestRenderer;
  private instances?: ReactTestInstance[];

  constructor(
    rootRenderer: ReactTestRenderer,
    instances?: ReactTestInstance[]
  ) {
    this.rootRenderer = rootRenderer;
    this.instances = instances;
  }

  toJSON() {
    return this.instances?.length
      ? toJSON(this.instances[0])
      : this.rootRenderer.toJSON();
  }

  find<T>(selector: React.ElementType<T>) {
    return new ReactWrapper(
      this.rootRenderer,
      this.rootRenderer.root.findAllByType(selector)
    );
  }

  exists() {
    return (this.instances?.length ?? 0) > 0;
  }
}

export function mount(element: ReactElement) {
  return new ReactWrapper(create(element));
}
