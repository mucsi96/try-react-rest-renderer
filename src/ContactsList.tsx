import { FC, useEffect, useState } from "react";
import { ContactRow } from "./ContactRow";

async function getContacts() {
  await new Promise((resolve) => setTimeout(resolve, 10));
  return ["Tom", "Alex", "Viki"];
}

export const ContactsList: FC = () => {
  const [contacts, setContacts] = useState<string[]>();

  useEffect(() => {
    getContacts().then(setContacts);
  }, []);

  return (
    <ul>
      {contacts?.map((contact) => (
        <li key={contact}>
          <ContactRow>{contact}</ContactRow>
        </li>
      ))}
    </ul>
  );
};
