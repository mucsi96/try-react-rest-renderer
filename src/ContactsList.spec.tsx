import { describe, expect, test, vi } from "vitest";
import { ContactRow } from "./ContactRow";
import { ContactsList } from "./ContactsList";
import { mount, waitFor } from "./testUtils";

vi.mock("./ContactRow", () => ({ ContactRow: "ContactRow-" }));

describe("ContactsList", () => {
  test("renders", async () => {
    const wrapper = mount(<ContactsList />);
    await waitFor(() => {
      expect(wrapper.find(ContactRow).exists()).toBe(true);
    });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ContactRow)).toMatchSnapshot();
  });
});
