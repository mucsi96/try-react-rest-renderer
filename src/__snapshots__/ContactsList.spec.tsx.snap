// Vitest Snapshot v1

exports[`ContactsList > renders 1`] = `
<ul>
  <li>
    <ContactRow->
      Tom
    </ContactRow->
  </li>
  <li>
    <ContactRow->
      Alex
    </ContactRow->
  </li>
  <li>
    <ContactRow->
      Viki
    </ContactRow->
  </li>
</ul>
`;

exports[`ContactsList > renders 2`] = `
<ContactRow->
  Tom
</ContactRow->
`;
